infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogReplicationSinkManager.java

```diff
@@ -318,16 +357,38 @@ private void applySnapshotSync(LogReplicationEntry message) {
             case SNAPSHOT_MESSAGE:
                 snapshotWriter.apply(message);
                 return;
-            case SNAPSHOT_END:
-                snapshotWriter.snapshotTransferDone(message);
-                completeSnapshotApply(message);
+
+            case SNAPSHOT_TRANSFER_END:
+                // Submit a job to the executor
+                applySnapshotExecutor.submit(() -> processSnapshotTransferEndMarker(message));
+
+                // Will trigger an ACK for SNAPSHOT_END MARKER
                 return;
+
             default:
                 log.warn("Message type {} should not be applied as snapshot sync.", message.getMetadata().getMessageMetadataType());
                 break;
         }
     }
 
+    /**
+     * Query metadata corfu table and get the most recent information.
+     * @return
+     */
+    public LogReplicationQueryMetadataResponse processQueryMetadataRequest() {
+        CorfuStoreMetadata.Timestamp timestamp = logReplicationMetadataManager.getTimestamp();
+
+        LogReplicationQueryMetadataResponse response = new LogReplicationQueryMetadataResponse(
+                logReplicationMetadataManager.getTopologyConfigId(timestamp),
+                logReplicationMetadataManager.getVersion(timestamp),
```

Comment:
I am a bit concerned of the number of times we go to fetch separate values of the metadata. Can we at least group these 6 DB accesses?

Reply:
I am planning to address it in another PR. Change the formate of storing them as one meta key, and the value contains all the current keys.